src/compiler/crystal/syntax/parser.cr

```diff
@@ -3909,6 +3928,11 @@ module Crystal
       end_location = token_end_location
       doc = @token.doc
 
+      if @token.type == :"!"
+        obj = Var.new("self").at(location)
+        return parse_negation_suffix(obj)
+      end
```

Comment:
I don't think this is needed.

This would turn `!` into `!self`, but you never want to do that.

In other cases plain `is_a?` it turned into `self.is_a?`, but for `!` it doesn't make sense.

Maybe remove it and check that no parser spec is broken?

Reply:
It's currently takes effect at `parse_when_expression`, it breaks if removed.
On other places `parse_var_or_call` is not called if token is :!. So i can leave it, or make extra condition in `parse_when_expression` which can be a little bit messy.

Reply:
Ah, I see. Maybe add a comment saying that this will only trigger from `parse_when_expression`?

Reply:
yes, added.